core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java

```diff
@@ -85,32 +88,47 @@ private boolean checkDimensions(LPTableau tableau, Matrix<DRes<SInt>> updateMatr
   public DRes<LPOutput> buildComputation(ProtocolBuilderNumeric builder) {
     DRes<SInt> zero = builder.numeric().known(BigInteger.ZERO);
     return builder.seq(seq -> {
-      this.iterations = 0;
       List<BigInteger> enumeratedVariables = new ArrayList<>(noVariables);
       for (int i = 1; i <= noVariables; i++) {
         enumeratedVariables.add(BigInteger.valueOf(i));
       }
 
-      LpState initialState = new LpState(BigInteger.ZERO, tableau, updateMatrix, null, pivot,
+      LpState initialState = new LpState(tableau, updateMatrix, pivot,
           enumeratedVariables, initialBasis, pivot);
       return () -> initialState;
     }).whileLoop(state -> !state.terminated(), (seq, state) -> {
-      iterations++;
       if (isDebug()) {
         debugInfo(seq, state);
       }
       return seq.seq((inner) -> {
-        logger.info("LP Iterations=" + iterations + " solving " + identityHashCode);
+        logger.info("LP Iterations=" + state.iteration + " solving " + identityHashCode);
+        if (state.iteration >= maxNumberOfIterations) {
+          logger.info("Aborting " + identityHashCode + " no solution found");
+          return Pair.lazy(null, () -> BigInteger.ONE);
+        }
         if (pivotRule == PivotRule.BLAND) {
           return blandPhaseOneProtocol(inner, state);
         } else {
           return phaseOneProtocol(inner, state, zero);
         }
       }).seq((inner, phaseOneOutput) -> {
-        if (!phaseOneOutput.terminated()) {
-          phaseTwoProtocol(inner, phaseOneOutput);
+        BigInteger isFinished = phaseOneOutput.getSecond().out();
+        if (isFinished.equals(BigInteger.ONE)) {
```

Comment:
This sequence of if statements seems overly complex. Maybe it would help to use, say, BigInteger.TWO to signal the abort state?